File: Python/GenericStructures/Tree.py
from typing import TypeVar, Generic

T = TypeVar("T")


class Node(Generic[T]):
    left: Node[T] = None
    right: Node[T] = None

    def __init__(self, data: T = None):
        self.data = data


class Tree(Generic[T]):
    m_size: int = 0
    m_root: Node[T] = None

    def insert_node(self, node: Node[T]) -> None:
        if self.is_empty():
            self.m_root = node
        else:
            ptr = self.m_root
            while ptr != None:
                if node.data < ptr.data:
                    ptr = ptr.left
                elif node.data > ptr.data:
                    ptr = ptr.right
                else:
                    return

            if node.data < ptr.data:
                ptr.left = node
            if node.data > ptr.data:
                ptr.right = node

        self.m_size += 1

    def insert_element(self, data: int) -> None:
        node = Node[T](data)
        if self.is_empty():
            self.m_root = node
        else:
            ptr = self.m_root
            while ptr.left is not None or ptr.right is not None:
                if node.data < ptr.data:
                    ptr = ptr.left
                elif node.data > ptr.data:
                    ptr = ptr.right
                else:
                    return

            if node.data < ptr.data:
                ptr.left = node
            if node.data > ptr.data:
                ptr.right = node

        self.m_size += 1

    def traverse_inorder(self, node: Node[T]) -> None:
        if node is not None:
            self.traverse_inorder(node.left)
            print(node.data)
            self.traverse_inorder(node.right)

    def traverse_preorder(self, node: Node[T]) -> None:
        if node is not None:
            print(node.data)
            self.traverse_preorder(node.left)
            self.traverse_preorder(node.right)

    def traverse_postorder(self, node: Node[T]) -> None:
        if node is not None:
            self.traverse_postorder(node.left)
            self.traverse_postorder(node.right)
            print(node.data)

    def delete_element(self, node: Node[T], data: int) -> None:
        if node is not None:
            return node
        elif data < node.data:
            node.left = self.delete_element(node.left, data)
        elif data > node.data:
            node.right = self.delete_element(node.right, data)
        else:
            # case 1:
            if node.left is None and node.right is None:
                ptr = None
            # case 2:
            elif node.left is None:
                ptr = node
                node = node.right
                ptr = None

            elif node.right is None:
                ptr = node
                node = node.left
                ptr = None
            # case 3:
            else:
                ptr = self.find_min(node)
                node.data = ptr.data
                ptr.right = self.delete_element(node.right, ptr.data)

    def find_min(self, node: Node[T]) -> Node[T]:
        while node.left is not None:
            node = node.left
        return node

    def is_empty(self) -> bool:
        return self.m_root == None


if __name__ == "__main__":
    t: Tree = Tree()
    ptr: Node[T] = t.m_root

    for i in range(0, 10):
        t.insert_element(i)

    t.traverse_postorder(t.m_root)
    print("done")

File: Java/GenericStructures/Tree.java
package GenericStructures;

public class Tree<T extends Comparable<T>>{
    class Node{
        T data;
        Node left = null;
        Node right = null;
        public Node(T data){
            this.data = data;
        }
        public Node(){}
    }

    public Node root = null;
    public Tree(){}
    
    public void displayInorder(Node ptr){
        if(ptr!=null){
            displayInorder(ptr.left);
            System.out.println(ptr.data);
            displayInorder(ptr.right);
        }
    }

    public void displayPreorder(Node ptr){
        if(ptr!=null){
            System.out.println(ptr.data);
            displayPreorder(ptr.left);
            displayPreorder(ptr.right);
        }
    }

    public void displayPostorder(Node ptr){
        if(ptr!=null){
            displayPostorder(ptr.left);
            displayPostorder(ptr.right);
            System.out.println(ptr.data);
        }
    }

    public void insertNode(Node node){
        insertNodeElement(node.data);
    }

    public void insertNodeElement(T data){
        Node node = new Node(data);
        if(is_empty()){
            root = node;
        }else{
            Node ptr = root;
            while(ptr.left != null || ptr.right != null){
                if((node.data).compareTo(ptr.data) == -1)
                    ptr = ptr.left;
                if((node.data).compareTo(ptr.data) == 1)
                    ptr = ptr.right;
                if((node.data).compareTo(ptr.data) == 0)
                    return; // no duplicate
            }

            if((node.data).compareTo(ptr.data) == 1)
                ptr.right = node;
            else
                ptr.left = node;
        }
    }

    public Node deleteNode(Node ptr, T data){
        if(ptr==null)
            return ptr;
        else if((data).compareTo(ptr.data) == -1)
            ptr.left = deleteNode(ptr.left, data);
        else if(data.compareTo(ptr.data) == 1)
            ptr.right = deleteNode(ptr.right, data);
        else{
            // case 1: leaf node
            if(ptr.left == null && ptr.right == null){
                ptr = null;
            }
            // case 2: one child
            else if(ptr.left == null){
                Node temp = ptr;
                ptr = ptr.right;
                temp = null;
            }
            else if(ptr.right == null){
                Node temp = ptr;
                ptr = ptr.left;
                temp = null;
            }
            // case 3: 2 children
            else{
                Node temp  = findMin(ptr.right);
                ptr.data = temp.data;
                ptr.right = deleteNode(ptr.right, temp.data);
            }

            return ptr;
        }

    }

    public Node findMin(Node ptr){
        while(ptr.left != null)
            ptr = ptr.left;
        return ptr;
    }

    public Node findMax(Node ptr){
        while(ptr.right != null)
            ptr = ptr.right;
        return ptr;
    }

    public boolean is_element(T data){
        Node ptr = root;
        while(ptr.left != null || ptr.right != null){
            if((data).compareTo(ptr.data) == -1)
                ptr = ptr.left;
            if(data.compareTo(ptr.data) == 1)
                ptr = ptr.right;
            if(data.compareTo(ptr.data) == 0)
                return true; 
        }
        return false;
    }

    private boolean is_empty(){
        return root == null;
    }
}

class MainClass{
    public static void main(String[] args){
        Tree<Integer> tree = new Tree<>();

        for(int i = 0; i < 10; ++i){
            tree.insertNodeElement(i);
        }
        tree.displayInorder(tree.root);
    }
}
